sale_product_kit/models/product_template.py

```diff
@@ -0,0 +1,45 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    is_kit = fields.Boolean(string="Is Kit")
+
+    sub_product_ids = fields.Many2many(
+        comodel_name='product.product',
+        string="Sub Products",
+        help="Products that are part of this kit.",
+        domain=[
+            ('sale_ok', '=', True),
+            ('type', '!=', 'combo'),
+            ('is_kit', '!=', True),
+        ],
+    )
+
+    @api.constrains('is_kit')
+    def _check_not_sub_product(self):
+        for template in self:
+            if template.is_kit:
+                kit_products = self.env['product.template'].search([
+                    '&',
+                    ('is_kit', '=', True),
+                    ('sub_product_ids', 'in', template.product_variant_ids.ids),
+                ])
+                if kit_products:
+                    raise ValidationError(_("This product cannot become a kit since it is a sub-product of another kit."))
+
+    @api.constrains('is_kit', 'sub_product_ids')
+    def _check_sub_product_ids_not_empty(self):
+        for template in self:
+            if template.is_kit and not template.sub_product_ids:
+                raise ValidationError(_("A kit product must contain at least 1 sub product."))
```

Comment:
Similarly for this, instead of iterating for each product, we can use `.filtered()`.
```suggestion
    def _check_sub_product_ids_not_empty(self):
        if self.filtered(lambda p: p.is_kit and not p.sub_product_ids.exists()):
            raise ValidationError(_("A kit product must contain at least one sub-product."))
```

Reply:
In my updated code, I used any() as i found that more clear to read than filtered().